connectors/sources/salesforce.py

```diff
@@ -169,26 +206,118 @@ async def _select_queryable_fields(self, sobject, fields):
+                    break
+                else:
+                    next_url = next_page.get("nextRecordsUrl")
 
     def _token_cleanup(self):
         self.token = None
-        self.token_issued_at = None
-        self.token_refresh_enabled = False
 
     def _auth_headers(self):
         return {"authorization": f"Bearer {self.token}"}
 
-    async def get_request(self, endpoint, params=None):
-        # TODO improve error handling
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+    )
+    async def _get_json(self, url, params=None):
+        response_body = None
+
+        try:
+            response = await self._get(url, params=params)
+            response_body = await response.json()
+            # We get the response body before raising for status as it contains vital error information
+            response.raise_for_status()
+            return response_body
+        except ClientResponseError as e:
+            await self._handle_client_response_error(response_body, e)
+        except Exception as e:
+            raise e
+
+    async def _get(self, url, params=None):
         return await self.session.get(
-            f"{self.base_url}{endpoint}", headers=self._auth_headers(), params=params
+            url,
+            headers=self._auth_headers(),
+            params=params,
         )
 
+    async def _post(self, url, data=None):
+        return await self.session.post(url, data=data)
+
+    async def _handle_client_response_error(self, response_body, e):
+        exception_details = f"status: {e.status}, message: {e.message}"
+
+        if e.status == 401:
+            self._logger.warning(
+                f"Token expired, attemping to fetch new token. Status: {e.status}, message: {e.message}"
+            )
+            # The user can alter the lifetime of issued tokens, so we don't know when they expire
+            # Therefore we fetch the token when we encounter an error rather than when it expires
+            await self.get_token()
+            # raise to continue with retry strategy
+            raise e
+        elif 400 <= e.status < 500:
+            errors = self._handle_response_body_error(response_body)
+            # errorCode and message are generally identical, except if the query is invalid
+            error_codes = [x["errorCode"] for x in errors]
+
+            if "REQUEST_LIMIT_EXCEEDED" in error_codes:
+                raise RateLimitedException(
+                    f"Salesforce is rate limiting this account. {exception_details}, details: {', '.join(error_codes)}"
+                ) from e
+            elif (
+                any(
+                    error in error_codes
+                    for error in [
+                        "INVALID_FIELD",
+                        "INVALID_TERM",
+                        "MALFORMED_QUERY",
+                    ]
+                )
+                in error_codes
+            ):
+                raise InvalidQueryException(
```

Comment:
If it's `RateLimitedException` or `InvalidQueryException`, we should not retry

Reply:
You can use this when it's merged: https://github.com/elastic/connectors-python/pull/1395

Reply:
That was a really convenient update, thank you for adding it!